pkg/leader/leader.go

```diff
@@ -0,0 +1,194 @@
+		return nil
+	}
+	return err
+}
+
+// Become ensures that the current pod is the leader within its namespace. It
+// continuously tries to create a ConfigMap with the provided name and the
+// current pod set as the owner reference. Only one can exist at a time with
+// the same name, so the pod that successfully creates the ConfigMap is the
+// leader. Upon termination of that pod, the garbage collector will delete the
+// ConfigMap, enabling a different pod to become the leader.
+func Become(ctx context.Context, lockName string) error {
+	logrus.Info("trying to become the leader")
+
+	ns, err := myNS()
+	if err != nil {
+		return err
+	}
+
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return err
+	}
+
+	client, err := crclient.New(config, crclient.Options{})
+	if err != nil {
+		return err
+	}
+
+	owner, err := myOwnerRef(ctx, client, ns)
+	if err != nil {
+		return err
+	}
+
+	// check for existing lock from this pod, in case we got restarted
+	existing := &corev1.ConfigMap{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "ConfigMap",
+		},
+	}
+	key := crclient.ObjectKey{Namespace: ns, Name: lockName}
+	err = client.Get(ctx, key, existing)
+
+	switch {
+	case err == nil:
+		for _, existingOwner := range existing.GetOwnerReferences() {
+			if existingOwner.Name == owner.Name {
+				logrus.Info("Found existing lock with my name. I was likely restarted.")
+				logrus.Info("Continuing as the leader.")
+				return nil
+			} else {
+				logrus.Infof("Found existing lock from %s", existingOwner.Name)
+			}
+		}
+	case apierrors.IsNotFound(err):
+		logrus.Info("No pre-existing lock was found.")
+	default:
+		logrus.Error("unknown error trying to get ConfigMap")
+		return err
+	}
+
+	cm := &corev1.ConfigMap{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "ConfigMap",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            lockName,
+			Namespace:       ns,
+			OwnerReferences: []metav1.OwnerReference{*owner},
+		},
+	}
+
+	// try to create a lock
+	backoff := time.Second
+	for {
+		err := client.Create(ctx, cm)
+		switch {
+		case err == nil:
+			logrus.Info("Became the leader.")
+			return nil
+		case apierrors.IsAlreadyExists(err):
+			logrus.Info("Not the leader. Waiting.")
+			select {
+			case <-time.After(wait.Jitter(backoff, .2)):
+				if backoff < maxBackoffInterval {
+					backoff *= 2
+				}
+				continue
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+		default:
+			logrus.Error("unknown error creating configmap")
+			return err
+		}
+	}
+}
+
+// myNS returns the name of the namespace in which this code is currently running.
+func myNS() (string, error) {
+	nsBytes, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace")
+	if err != nil {
+		if os.IsNotExist(err) {
+			return "", ErrNoNS
+		}
+		return "", err
+	}
+	ns := strings.TrimSpace(string(nsBytes))
+	logrus.Infof("found namespace: %s", ns)
```

Comment:
Probably no need to log this. It's not saying much as this is normal behavior and users already know what namespace they're running the operator in.

Reply:
Good point. I'll change it to Debug level since it could be useful for troubleshooting.